iris_grib/tests/unit/save_rules/test_grid_definition_template_4.py

```diff
@@ -0,0 +1,99 @@
+# the terms of the GNU Lesser General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# iris-grib is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with iris-grib.  If not, see <http://www.gnu.org/licenses/>.
+"""
+Unit tests for :func:`iris_grib._save_rules.grid_definition_template_4`.
+
+"""
+
+from __future__ import (absolute_import, division, print_function)
+from six.moves import (filter, input, map, range, zip)  # noqa
+
+# Import iris_grib.tests first so that some things can be initialised before
+# importing anything else.
+import iris_grib.tests as tests
+
+import numpy as np
+
+from iris.coord_systems import GeogCS
+
+from iris_grib._save_rules import grid_definition_template_4
+from iris_grib.tests.unit.save_rules import GdtTestMixin
+
+
+class Test(tests.IrisGribTest, GdtTestMixin):
+    def setUp(self):
+        GdtTestMixin.setUp(self)
+
+    def _default_x_points(self):
+        # Define an irregular x-coord, to force template 3.5 instead of 3.1.
+        return [1.0, 2.0, 5.0]
+
+    def test__template_number(self):
+        grid_definition_template_4(self.test_cube, self.mock_grib)
+        self._check_key('gridDefinitionTemplateNumber', 4)
+
+    def test__shape_of_earth_spherical(self):
+        cs = GeogCS(semi_major_axis=1.23)
+        test_cube = self._make_test_cube(cs=cs)
+        grid_definition_template_4(test_cube, self.mock_grib)
+        self._check_key('shapeOfTheEarth', 1)
+        self._check_key('scaleFactorOfRadiusOfSphericalEarth', 0)
+        self._check_key('scaledValueOfRadiusOfSphericalEarth', 1.23)
+
+    def test__shape_of_earth_flattened(self):
+        cs = GeogCS(semi_major_axis=1.456,
+                    semi_minor_axis=1.123)
+        test_cube = self._make_test_cube(cs=cs)
+        grid_definition_template_4(test_cube, self.mock_grib)
+        self._check_key('shapeOfTheEarth', 7)
+        self._check_key('scaleFactorOfEarthMajorAxis', 0)
+        self._check_key('scaledValueOfEarthMajorAxis', 1.456)
+        self._check_key('scaleFactorOfEarthMinorAxis', 0)
+        self._check_key('scaledValueOfEarthMinorAxis', 1.123)
+
+    def test__grid_shape(self):
+        test_cube = self._make_test_cube(x_points=np.arange(13),
+                                         y_points=np.arange(6))
+        grid_definition_template_4(test_cube, self.mock_grib)
+        self._check_key('Ni', 13)
+        self._check_key('Nj', 6)
+
+    def test__grid_points(self):
+        x_floats = np.array([11.0, 12.0, 167.0])
+        # TODO: reduce Y to 2 points, when gribapi nx=ny limitation is gone.
```

Comment:
Wow, is that really a thing?  Who thought that would be a good idea?!

Reply:
The TODO says it is (or it was)... I haven't investigated it further as I don't think it impacts these tests at all.

Reply:
> I haven't investigated it further as I don't think it impacts these tests

Ooh :+1: for work-not-done ! :mortar_board: